### PR TITLE
 Extend NDISAPI with Asynchronous Functionality

### PR DESCRIPTION
**Pull request by wiresock**
This PR introduces asynchronous extensions to the NDISAPI. The primary aim of these extensions is to provide non-blocking I/O operations, enhancing the library's performance by leveraging Rust's async/await syntax and the futures library.
In order to demonstrate the newly added async functionalities, two sample applications, `async_packthru.rs` and `async_passthru.rs`, have been included in this commit. These samples showcase the practical usage and benefits of the asynchronous extensions in real-world scenarios.

### CODE REVIEW

File: examples/listadapters.rs [hunk 1]
```diff
@@ -26,7 +26,7 @@ fn main() -> Result<()> {
     for (index, adapter) in adapters.iter().enumerate() {
         // Display the information about each network interface provided by the get_tcpip_bound_adapters_info
         let network_interface_name = Ndisapi::get_friendly_adapter_name(adapter.get_name())
-            .expect("Unkown network interface");
+            .expect("Unknown network interface");
```
Comment (jamilbk) [inline examples/listadapters.rs:29]: We use `codespell` (with pre-commit) in to catch typos like this in the `firezone/firezone` repo. If it's helpful I could add it to the CI here?

**Review by jamilbk — COMMENTED** [review]
I'll defer to @conectado for in-depth review but I a minor suggestion. LGTM, thanks for adding the notes for the `unsafe` blocks.

File: Cargo.toml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "ndisapi"
-version = "0.2.2"
+version = "0.3.0"
```
Comment (jamilbk) [inline Cargo.toml:3]: You may also want to update the [associated README instructions](https://github.com/firezone/ndisapi/blob/main/README.md?plain=1#L31) to use `0.3.0`

Comment (jamilbk) [reply Cargo.toml:3]: Edit: nevermind, I see that you already did. Need more coffee 😅

File: examples/listadapters.rs (hunk 1 above)
Comment (wiresock) [reply examples/listadapters.rs:29]: Yes, please. An additional check is always beneficial. This is especially true for open source projects, where I need to exercise extra caution when it comes to code style, identifying typos, and ensuring proper formatting.

Comment (jamilbk) [reply examples/listadapters.rs:29]: Tracking in https://github.com/firezone/firezone/issues/1634

**Pull request closed without merging**